### PR TITLE
implement dlerror & dlclose

### PR DESCRIPTION
**Pull request by TotallyGamerJet**

### CODE REVIEW

File: dlfcn.go [hunk 1]
```diff
@@ -17,20 +18,66 @@ const RTLD_GLOBAL = 0x8
...
+		// use unsafe.Add once we reach 1.17
+		//_ = (*byte)(unsafe.Add(unsafe.Pointer(msg), length))
+		if *(*byte)(unsafe.Pointer(msg + uintptr(length))) == '\x00' {
+			break
+		}
+		length++
+	}
+	// use unsafe.Slice once we reach 1.17
+	// we make a slice then cast to a string to force Go to copy the string from C memory into Go memory
+	return string(*(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{Data: msg, Len: length, Cap: length})))
```
Comment (hajimehoshi) [inline dlfcn.go:67]: What is the lifetime of msg?

**Review by hajimehoshi — COMMENTED** [review]
I think it is better to split the PR into the documentation part and the Dlerror part.

File: dlfcn.go (hunk 1 above)
Comment (TotallyGamerJet) [reply dlfcn.go:67]: It is valid until the next call to dlerror. Which is why we copy it into Go so that Go can decide when it no longer needs the string

Comment (hajimehoshi) [reply dlfcn.go:67]: Does this line really copy the string? Probably at casting from []byte to string?

Comment (TotallyGamerJet) [reply dlfcn.go:67]: Yeah a cast []byte to string copies bc it can’t prove that the []byte won’t be modified and break the immutability of string

**Comment by TotallyGamerJet** [comment]
I will split these once I have a chance

File: dlfcn.go (hunk 1 above)
Comment (TotallyGamerJet) [reply dlfcn.go]: I went ahead and made the copying more obvious

File: dlfcn.go [hunk 2]
```diff
@@ -31,6 +32,32 @@ func Dlsym(handle uintptr, name string) uintptr {
 	return ret
 }
 
+func Dlerror() string {
+	msg, _, _ := SyscallN(dlerrorABI0)
```
Comment (hajimehoshi) [inline dlfcn.go:36]: Add a comment about the lifetime of `msg`.

**Review by hajimehoshi — APPROVED** [review]
LGTM % one comment

**Comment by hajimehoshi** [comment]
So, are you sending a PR to add comments later? Thanks,

**Pull request merged**